File: src/api/sources/mangakatana.ts
// MangaKatana (mangakatana.com) source — a plain HTML site we scrape. Replaces
// the old mangadot stub. Chapter pages hide the real image list in a JS array
// (with decoy arrays alongside it), so we pick the longest one.

import { httpText } from "../client.ts";
import { readCache, writeCache } from "../../utils/cache.ts";
import { MANGA_CACHE, SEARCH_CACHE, CHAPTERS_CACHE } from "../../utils/paths.ts";
import type { DiscoveryItem, Filters, MangaInfo, ReadChapter, SearchResult, Source } from "../types.ts";

const BASE = "https://mangakatana.com";
const HEADERS = { Referer: BASE + "/" };
const TTL = 60 * 60 * 1000;

function decodeEntities(s: string): string {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#0?39;|&apos;/g, "'")
    .replace(/&#0?34;/g, '"')
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .trim();
}

/** Manga id = the path after /manga/, e.g. "berserk.1087". */
function idFromUrl(url: string): string {
  return url.replace(/^https?:\/\/mangakatana\.com\/manga\//, "").replace(/\/.*$/, "");
}

function parseSearchItems(html: string): SearchResult[] {
  // A unique exact match 302s straight to the manga page (which has an
  // <h1 class="heading">); return just that title rather than its "related" widget.
  const h1 = html.match(/<h1 class="heading">([^<]+)<\/h1>/);
  const canon = html.match(/<link rel="canonical" href="(https:\/\/mangakatana\.com\/manga\/[^"]+)"/);
  if (h1 && canon) {
    const cover = html.match(/<meta property="og:image" content="([^"]+)"/)?.[1];
    return [
      { id: idFromUrl(canon[1]), title: decodeEntities(h1[1]), poster: cover, type: "Manga", isAdult: false, source: "mangakatana" },
    ];
  }
  // Otherwise it's a results list: only the result cards carry `data-id`
  // (sidebar/latest widgets use a plain `<div class="item">`).
  const out: SearchResult[] = [];
  const seen = new Set<string>();
  for (const m of html.matchAll(
    /<div class="item" data-genre="[^"]*" data-id="\d+"([\s\S]*?)<h3 class="title">\s*<a[^>]*href="(https:\/\/mangakatana\.com\/manga\/[^"]+)"[^>]*>([^<]+)<\/a>/g,
  )) {
    const id = idFromUrl(m[2]);
    if (seen.has(id)) continue;
    seen.add(id);
    const cover = m[1].match(/<img[^>]+src="(https:\/\/mangakatana\.com\/imgs\/[^"]+)"/)?.[1];
    out.push({ id, title: decodeEntities(m[3]), poster: cover, type: "Manga", isAdult: false, source: "mangakatana" });
  }
  return out;
}

export const mangakatana: Source = {
  id: "mangakatana",
  label: "MangaKatana (mangakatana.com)",
  available: true,

  async search(query): Promise<SearchResult[]> {
    const key = `mk:search:${query}`;
    const hit = await readCache<SearchResult[]>(SEARCH_CACHE, key, 10 * 60 * 1000);
    if (hit) return hit;
    const html = await httpText(`${BASE}/?search=${encodeURIComponent(query)}&search_by=book_name`, { headers: HEADERS });
    const results = parseSearchItems(html).slice(0, 24);
    await writeCache(SEARCH_CACHE, key, results);
    return results;
  },

  // Discovery/genres come from other sources; MangaKatana is a search+read backend.
  async discovery(): Promise<DiscoveryItem[]> {
    return [];
  },
  async filters(): Promise<Filters> {
    return { genres: [], tags: [], types: [], statuses: [] };
  },
  async browse(): Promise<SearchResult[]> {
    return [];
  },

  async info(mangaId): Promise<MangaInfo> {
    const key = `mk:info:${mangaId}`;
    const hit = await readCache<MangaInfo>(MANGA_CACHE, key, TTL);
    if (hit) return hit;

    const page = await httpText(`${BASE}/manga/${mangaId}`, { headers: HEADERS });
    const title = decodeEntities(page.match(/<h1 class="heading">([^<]+)<\/h1>/)?.[1] ?? "Untitled");
    // The page is noisy (sidebars, recommendations), so we don't auto-detect
    // long-strip here — readers can toggle it with `w`.
    const longStrip = false;

    const raw: Array<{ seg: string; number: number }> = [];
    const seen = new Set<string>();
    for (const m of page.matchAll(
      /href="https:\/\/mangakatana\.com\/manga\/[^"]*\/(c[0-9.]+)"[^>]*>\s*(?:Chapter|Ch\.?|Episode)?\s*([0-9]+(?:\.[0-9]+)?)/gi,
    )) {
      const seg = m[1];
      if (seen.has(seg)) continue;
      seen.add(seg);
      raw.push({ seg, number: parseFloat(m[2]) });
    }
    raw.sort((a, b) => a.number - b.number);

    const info: MangaInfo = {
      id: mangaId,
      title,
      type: longStrip ? "Manhwa" : "Manga",
      forceStrip: longStrip,
      chapters: raw.map((ch, i) => ({
        id: ch.seg,
        title: `Chapter ${ch.number}`,
        number: ch.number,
        index: i,
        pageCount: 0,
        scanId: "",
      })),
    };
    await writeCache(MANGA_CACHE, key, info);
    return info;
  },

  async pages(mangaId, chapterSeg): Promise<ReadChapter> {
    const key = `mk:pages:${mangaId}/${chapterSeg}`;
    const hit = await readCache<ReadChapter>(CHAPTERS_CACHE, key, TTL);
    if (hit) return hit;
    const html = await httpText(`${BASE}/manga/${mangaId}/${chapterSeg}`, { headers: HEADERS });
    // The page declares several `var x=[ '…' ]` arrays; only one holds the real
    // images (the rest are decoys). Pick the longest array of image URLs.
    let best: string[] = [];
    for (const m of html.matchAll(/var\s+[a-zA-Z_$]+\s*=\s*\[([\s\S]*?)\]\s*;/g)) {
      const urls = [...m[1].matchAll(/["'](https?:\/\/[^"']+\.(?:jpg|jpeg|png|webp|avif)[^"']*)["']/gi)].map((x) => x[1]);
      if (urls.length > best.length) best = urls;
    }
    const out: ReadChapter = {
      id: chapterSeg,
      title: "",
      scanId: "",
      pages: best.map((url, i) => ({ id: `${chapterSeg}-${i}`, url, number: i, width: 0, height: 0, aspectRatio: 1 })),
    };
    await writeCache(CHAPTERS_CACHE, key, out);
    return out;
  },

  async related(): Promise<DiscoveryItem[]> {
    return [];
  },
};

File: go-rewrite/internal/api/mangakatana.go
// MangaKatana (mangakatana.com) source — a plain HTML site we scrape. Chapter
// pages hide the real image list in a JS array (with decoy arrays alongside),
// so we pick the longest one.

package api

import (
	"fmt"
	"html"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/devcheckra1n/manga-cli/go-rewrite/internal/util"
)

const (
	mkBase = "https://mangakatana.com"
	mkTTL  = time.Hour
)

var mkHeaders = map[string]string{"Referer": mkBase + "/"}

var (
	mkH1Re    = regexp.MustCompile(`<h1 class="heading">([^<]+)</h1>`)
	mkCanonRe = regexp.MustCompile(`<link rel="canonical" href="(https://mangakatana\.com/manga/[^"]+)"`)
	mkOgImgRe = regexp.MustCompile(`<meta property="og:image" content="([^"]+)"`)
	mkItemRe  = regexp.MustCompile(`(?s)<div class="item" data-genre="[^"]*" data-id="\d+"(.*?)<h3 class="title">\s*<a[^>]*href="(https://mangakatana\.com/manga/[^"]+)"[^>]*>([^<]+)</a>`)
	mkCoverRe = regexp.MustCompile(`<img[^>]+src="(https://mangakatana\.com/imgs/[^"]+)"`)
	mkChapRe  = regexp.MustCompile(`(?i)href="https://mangakatana\.com/manga/[^"]*/(c[0-9.]+)"[^>]*>\s*(?:Chapter|Ch\.?|Episode)?\s*([0-9]+(?:\.[0-9]+)?)`)
	mkArrayRe = regexp.MustCompile(`(?s)var\s+[a-zA-Z_$]+\s*=\s*\[(.*?)\]\s*;`)
	mkImgRe   = regexp.MustCompile(`(?i)["'](https?://[^"']+\.(?:jpg|jpeg|png|webp|avif)[^"']*)["']`)
)

type mangakatanaSource struct{}

func (mangakatanaSource) ID() SourceID    { return SrcMangaKatana }
func (mangakatanaSource) Label() string   { return "MangaKatana (mangakatana.com)" }
func (mangakatanaSource) Available() bool { return true }

// mkIDFromURL: manga id = the path after /manga/, e.g. "berserk.1087".
func mkIDFromURL(raw string) string {
	s := strings.TrimPrefix(raw, "https://mangakatana.com/manga/")
	s = strings.TrimPrefix(s, "http://mangakatana.com/manga/")
	if i := strings.IndexByte(s, '/'); i >= 0 {
		s = s[:i]
	}
	return s
}

func mkParseSearch(doc string) []SearchResult {
	// A unique exact match 302s straight to the manga page (which has an
	// <h1 class="heading">); return just that title, not its "related" widget.
	h1 := mkH1Re.FindStringSubmatch(doc)
	canon := mkCanonRe.FindStringSubmatch(doc)
	if h1 != nil && canon != nil {
		cover := ""
		if m := mkOgImgRe.FindStringSubmatch(doc); m != nil {
			cover = m[1]
		}
		return []SearchResult{{
			ID:     mkIDFromURL(canon[1]),
			Title:  strings.TrimSpace(html.UnescapeString(h1[1])),
			Poster: cover,
			Type:   "Manga",
			Source: SrcMangaKatana,
		}}
	}
	// Otherwise a results list: only real result cards carry `data-id`.
	var out []SearchResult
	seen := map[string]bool{}
	for _, m := range mkItemRe.FindAllStringSubmatch(doc, -1) {
		id := mkIDFromURL(m[2])
		if seen[id] {
			continue
		}
		seen[id] = true
		cover := ""
		if c := mkCoverRe.FindStringSubmatch(m[1]); c != nil {
			cover = c[1]
		}
		out = append(out, SearchResult{
			ID: id, Title: strings.TrimSpace(html.UnescapeString(m[3])),
			Poster: cover, Type: "Manga", Source: SrcMangaKatana,
		})
	}
	return out
}

func (mangakatanaSource) Search(query string, _ SearchOpts) ([]SearchResult, error) {
	key := "mk:search:" + query
	var cached []SearchResult
	if util.ReadCache(util.SearchCache, key, 10*time.Minute, &cached) {
		return cached, nil
	}
	doc, err := httpText(mkBase+"/?search="+url.QueryEscape(query)+"&search_by=book_name", mkHeaders)
	if err != nil {
		return nil, err
	}
	results := mkParseSearch(doc)
	if len(results) > 24 {
		results = results[:24]
	}
	util.WriteCache(util.SearchCache, key, results)
	return results, nil
}

// Discovery/genres come from other sources; MangaKatana is a search+read backend.
func (mangakatanaSource) Discovery(DiscoveryKind, int, bool) ([]DiscoveryItem, error) {
	return []DiscoveryItem{}, nil
}
func (mangakatanaSource) Filters() (Filters, error) {
	return Filters{Genres: []Genre{}, Tags: []Genre{}, Types: []Genre{}, Statuses: []Genre{}}, nil
}
func (mangakatanaSource) Browse(BrowseFilter) ([]SearchResult, error) {
	return []SearchResult{}, nil
}

func (mangakatanaSource) Info(mangaID string) (*MangaInfo, error) {
	key := "mk:info:" + mangaID
	var cached MangaInfo
	if util.ReadCache(util.MangaCache, key, mkTTL, &cached) {
		return &cached, nil
	}

	doc, err := httpText(mkBase+"/manga/"+mangaID, mkHeaders)
	if err != nil {
		return nil, err
	}
	title := "Untitled"
	if m := mkH1Re.FindStringSubmatch(doc); m != nil {
		title = strings.TrimSpace(html.UnescapeString(m[1]))
	}
	// The page is noisy (sidebars, recommendations), so we don't auto-detect
	// long-strip here — readers can toggle it with `w`.

	type rawCh struct {
		seg string
		num float64
	}
	var raw []rawCh
	seen := map[string]bool{}
	for _, m := range mkChapRe.FindAllStringSubmatch(doc, -1) {
		if seen[m[1]] {
			continue
		}
		seen[m[1]] = true
		n, _ := strconv.ParseFloat(m[2], 64)
		raw = append(raw, rawCh{m[1], n})
	}
	sort.SliceStable(raw, func(a, b int) bool { return raw[a].num < raw[b].num })

	info := MangaInfo{ID: mangaID, Title: title, Type: "Manga", ForceStrip: false}
	for i, ch := range raw {
		info.Chapters = append(info.Chapters, Chapter{
			ID: ch.seg, Title: "Chapter " + trimFloat(ch.num), Number: ch.num, Index: i,
		})
	}
	util.WriteCache(util.MangaCache, key, info)
	return &info, nil
}

func (mangakatanaSource) Pages(mangaID, chapterSeg string) (*ReadChapter, error) {
	key := fmt.Sprintf("mk:pages:%s/%s", mangaID, chapterSeg)
	var cached ReadChapter
	if util.ReadCache(util.ChaptersCache, key, mkTTL, &cached) {
		return &cached, nil
	}
	doc, err := httpText(mkBase+"/manga/"+mangaID+"/"+chapterSeg, mkHeaders)
	if err != nil {
		return nil, err
	}
	// Several `var x=[ '…' ]` arrays exist; only one holds the real images
	// (the rest are decoys). Pick the longest array of image URLs.
	var best []string
	for _, m := range mkArrayRe.FindAllStringSubmatch(doc, -1) {
		var urls []string
		for _, im := range mkImgRe.FindAllStringSubmatch(m[1], -1) {
			urls = append(urls, im[1])
		}
		if len(urls) > len(best) {
			best = urls
		}
	}
	out := ReadChapter{ID: chapterSeg}
	for i, u := range best {
		out.Pages = append(out.Pages, Page{
			ID: fmt.Sprintf("%s-%d", chapterSeg, i), URL: u, Number: i, AspectRatio: 1,
		})
	}
	util.WriteCache(util.ChaptersCache, key, out)
	return &out, nil
}

func (mangakatanaSource) Related(string, int, bool) ([]DiscoveryItem, error) {
	return []DiscoveryItem{}, nil
}
